### PR TITLE
[DC] Adding `dc-print-dot` pass

### PR DESCRIPTION
**Pull request by luisacicolini**
This PR adds a new `dc-print-dot` pass that emulates `handshake-print-dot`. The main difference is that it produces dataflow graphs highlighting the separation between data and control, namely [DC values and tokens](https://circt.llvm.org/docs/Dialects/DC/).

### CODE REVIEW

File: include/circt/Dialect/DC/DCPasses.h [hunk 1]
```diff
@@ -28,4 +33,4 @@ std::unique_ptr<mlir::Pass> createDCDematerializeForksSinksPass();
 } // namespace dc
 } // namespace circt
 
-#endif // CIRCT_DIALECT_DC_DCPASSES_H
+#endif // CIRCT_DIALECT_DC_DCPASSES_H
```
Comment (TaoBi22) [inline include/circt/Dialect/DC/DCPasses.h:36]: Nit: newline at end of files

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    return DotNode{.nodeType = "pack",
+                   .incoming = valueToName(op.getOperands(), valuesMap, false),
+                   .outgoing = valueToName(op.getResults(), valuesMap, false)};
+  if (auto castOp = llvm::dyn_cast<dc::ToESIOp>(op))
+    return DotNode{.nodeType = "toESI",
+                   .incoming = valueToName(op.getOperands(), valuesMap, false),
+                   .outgoing = valueToName(op.getResults(), valuesMap, false)};
+  if (auto castOp = llvm::dyn_cast<dc::UnpackOp>(op))
+    return DotNode{.nodeType = "unpack",
+                   .incoming = valueToName(op.getOperands(), valuesMap, false),
```
Comment (TaoBi22) [inline lib/Dialect/DC/Transforms/Analysis.cpp:122]: Am I correct that the only thing that varies here between the DotNodes that are being returned is nodeType? It might be worth just letting these if statements define what nodeType is, and then feed that to just one call to the DotNode constructor which is probably a lot more concise

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      currentMap.push_back({v, name});
+    }
+  }
+  return res;
+}
+
+DotNode createDCNode(
+    Operation &op,
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> &valuesMap) {
+  if (auto castOp = llvm::dyn_cast<dc::ForkOp>(op))
```
Comment (TaoBi22) [inline lib/Dialect/DC/Transforms/Analysis.cpp:76]: Nit: castOp never seems to be used so you can just do
```suggestion
  if (isa<dc::ForkOp>(op))
```

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    os << "}";
+    outfile.close();
+  }
+};
+} // namespace
+
+std::unique_ptr<mlir::OperationPass<mlir::ModuleOp>>
+circt::dc::createDCDotPrintPass() {
+  return std::make_unique<DCDotPrintPass>();
+}
```
Comment (TaoBi22) [inline lib/Dialect/DC/Transforms/Analysis.cpp:283]: nit: newline at EOF

**Review by TaoBi22 — COMMENTED** [review]
Not super knowledgeable on DC so haven't looked in too much depth but just a few nits!

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    }
+  }
+  return res;
+}
+
+DotNode createDCNode(
+    Operation &op,
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> &valuesMap) {
+  if (auto castOp = llvm::dyn_cast<dc::ForkOp>(op))
+    return DotNode{.nodeType = "fork",
```
Comment (uenoku) [inline lib/Dialect/DC/Transforms/Analysis.cpp:77]: drive-by-comment: This designated initialization (https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) is c++ 20 feature so please avoid that. Also you might want to use `op.getName().stripDialect()` to get `nodeType` without hardcoding operation names here.

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+namespace {
+/// Emit the dot nodes
+struct DCDotPrintPass : public circt::dc::impl::DCDotPrintBase<DCDotPrintPass> {
+  void runOnOperation() override {
+
+    ModuleOp op = getOperation();
+
+    std::error_code ec;
+    llvm::raw_fd_ostream outfile("dc-out.dot", ec);
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/Analysis.cpp:173]: It might be beneficial when this file name could be provided as a pass parameter

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    mlir::raw_indented_ostream os(outfile);
+
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> valuesMap;
+    llvm::SmallVector<DotNode> nodes;
+
+    auto &moduleOps = op->getRegion(0).getBlocks();
+    // getBlock()->getOps<hw::HWModuleOp>();
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/Analysis.cpp:180]: Nit: Debug remnant?

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    tokenFlag = true;
+
+  DotNode n = {op.getName().stripDialect().str(),
+               valueToName(op.getOperands(), valuesMap, false),
+               valueToName(op.getOperands(), valuesMap, tokenFlag)};
+
+  return n;
+}
+
+DotNode createCombNode(
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/Analysis.cpp:88]: Nit: A comment for this function would be nice

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+using ValueMap = llvm::ScopedHashTable<mlir::Value, std::string>;
+
+// stores operand and results for each node in the dot graph
+struct DotNode {
+  std::string nodeType;
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/Analysis.cpp:43]: Nit: It might be a bit nicer and more efficient to use an enum for this instead. This then ofc. requires some stringify function.

File: lib/Dialect/DC/Transforms/Analysis.cpp (hunk 4 above)
Comment (Dinistro) [reply lib/Dialect/DC/Transforms/Analysis.cpp:238]: This was not resolved. You might have some auto format setting that reverts this change. IIRC, VS code has some odd behavior sometimes.

File: lib/Dialect/DC/Transforms/Analysis.cpp
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/Analysis.cpp:1]: I would suggest to rename this to `DCPrintDot.cpp` or something along these lines.

**Review by Dinistro — COMMENTED** [review]
Nice work :) I dropped a few nit comments mainly regarding style. I'm aware that the handshake pendant also violates many of the comments I've added, due to it being quite a relic.

I think this should probably be reviewed by @mortbopet .

**Comment by luisacicolini** [comment]
thanks a lot for your review! I fixed most minor comments and will fix the remaining ones as soon as possible :)

File: lib/Dialect/DC/Transforms/Analysis.cpp (hunk 6 above)
Comment (luisacicolini) [reply lib/Dialect/DC/Transforms/Analysis.cpp]: Thank you! Would it be better to pass it as a parameter or just print to std out (as for example [HW](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/Transforms/HWPrintInstanceGraph.cpp) does)?

File: lib/Dialect/DC/Transforms/Analysis.cpp (hunk 9 above)
Comment (Dinistro) [reply lib/Dialect/DC/Transforms/Analysis.cpp]: That's what I had in mind, yes.

File: llvm
Comment (Dinistro) [inline llvm:1]: You commited a submodule bump. This is probably not intended and might explain the CI failures.

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+using ValueMap = llvm::ScopedHashTable<mlir::Value, std::string>;
+
+namespace {
+
+/// all Comb and DC nodetypes
+enum NodeType {
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:41]:
```suggestion
enum class NodeType {
```
C enums should be avoided as they allow some weird conversions and do pollute namespaces. Note that changing this to `enum class` will force you to fully qualify later usages with `NodeType::`.

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+  addOp,
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:42]: Nit: We usually start enum members with a capital letter, similar to classes.
```suggestion
  AddOp,
```

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+  case sourceOp:
+    return "source";
+  case toESIOp:
+    return "toESI";
+  case unpackOp:
+    return "unpack";
+  case null:
+    return "null";
+  default:
+    return "Unknown NodeType";
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:122]:
```suggestion
```
Not adding a default type is preferred, as this ensures that you get a compiler warning when an enum member is not covered.

File: lib/Dialect/DC/Transforms/Analysis.cpp (hunk 6 above)
Comment (Dinistro) [reply lib/Dialect/DC/Transforms/Analysis.cpp]: Probably pass it as a parameter.

File: lib/Dialect/DC/Transforms/Analysis.cpp [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+  llvm::SmallVector<std::pair<mlir::Value, std::string>> incoming;
+  llvm::SmallVector<std::pair<mlir::Value, std::string>> outgoing;
+};
+
+// gives unique name to each value in the graph
+llvm::SmallVector<std::pair<mlir::Value, std::string>>
```
[the remark this answers is not in the document]
Comment (Dinistro) [reply lib/Dialect/DC/Transforms/Analysis.cpp]: Make sure to cover all such occurrences by grepping for them.

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+
+/// gives a unique name to each value in the graph
+llvm::SmallVector<std::pair<mlir::Value, std::string>> static valueToName(
+    const llvm::SmallVector<mlir::Value> &values,
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> &currentMap,
+    bool tokenFlag) {
+  SmallVector<std::pair<mlir::Value, std::string>> res;
+  for (auto [i, v] : llvm::enumerate(values)) {
+    auto found = false;
+    for (const auto &cm : currentMap) {
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:142]: Nit: you can probably use structured bindings as well here.
```suggestion
    for (const [key, value] &cm : currentMap) {
```

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+        }
+        }
+      });
+  return n;
+}
+
+namespace {
+/// Emit the dot nodes
+struct DCDotPrintPass : public circt::dc::impl::DCDotPrintBase<DCDotPrintPass> {
+  DCDotPrintPass(llvm::raw_ostream &os) : os(os) {}
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:269]:
```suggestion
  explicit DCDotPrintPass(llvm::raw_ostream &os) : os(os) {}
```
Nit: Without this, C++ allow to use this constructor for implicit type conversion. Fantastic language design, btw ;P

**Review by Dinistro — COMMENTED** [review]
Thanks for addressing the comments, dropped a few more nits, but that should be ready soonish 🙂

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 16]
```diff
@@ -0,0 +1,346 @@
+//===- DCPrintDot.cpp - Analysis Pass -----------------------------*- C++
+//-*-===//
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:2]: Nit: Make sure that this doesn't get wrapped. Same for other banners like this.

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Analysis pass.
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:10]: Nit: Outdated comment.

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+using namespace dc;
+using namespace mlir;
+
+using ValueMap = llvm::ScopedHashTable<mlir::Value, std::string>;
+
+namespace {
+
+/// all Comb and DC nodetypes
+class NodeType {
+  public:
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:42]: Sorry for the confusion, I was referring to an enum class like this:

```
enum class NodeType {
  AddOp,
  AndOp,
  ...
}
```

**Review by Dinistro — COMMENTED** [review]
Few more comments, then this will be ready to be shipped. Morten isn't around so I can accept it afterwards.

**Comment by Dinistro** [comment]
something with the submidule is still broken, btw

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp (hunk 18 above)
Comment (luisacicolini) [reply lib/Dialect/DC/Transforms/DCPrintDot.cpp:42]: oh sorry! got it now, thanks

**Comment by luisacicolini** [comment]
thanks! llvm is now back - now fixing builod issues

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+        }
+        }
+      });
+  return n;
+}
+
+namespace {
+/// Emit the dot nodes
+struct DCDotPrintPass : public circt::dc::impl::DCDotPrintBase<DCDotPrintPass> {
+  DCDotPrintPass(raw_ostream &os) : os(os) {}
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:266]:
```suggestion
  explicit DCDotPrintPass(raw_ostream &os) : os(os) {}
```

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 20]
```diff
@@ -0,0 +1,335 @@
...
+/// stores operand and results for each node in the dot graph
+struct DotNode {
+  NodeType nodeType;
+  SmallVector<std::pair<mlir::Value, std::string>> incoming;
+  SmallVector<std::pair<mlir::Value, std::string>> outgoing;
+};
+} // namespace
+
+/// gives a unique name to each value in the graph
+SmallVector<std::pair<mlir::Value, std::string>> static valueToName(
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:132]:
```suggestion
static SmallVector<std::pair<mlir::Value, std::string>> static valueToName(
```

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 21]
```diff
@@ -0,0 +1,335 @@
...
+        name = "token_" + std::to_string(currentMap.size());
+      res.push_back({v, name});
+      currentMap.push_back({v, name});
+    }
+  }
+  return res;
+}
+
+/// creates node in the dataflow graph for DC operations
+DotNode
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:157]:
```suggestion
static DotNode
```

File: lib/Dialect/DC/Transforms/DCPrintDot.cpp [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+      .Case([&](dc::SinkOp) { n.nodeType = NodeType::SinkOp; })
+      .Case([&](dc::SourceOp) { n.nodeType = NodeType::SourceOp; })
+      .Case([&](dc::ToESIOp) { n.nodeType = NodeType::ToESIOp; })
+      .Case([&](dc::UnpackOp) { n.nodeType = NodeType::UnpackOp; });
+
+  return n;
+}
+
+/// creates node in the dataflow graph for Comb operations
+DotNode
```
Comment (Dinistro) [inline lib/Dialect/DC/Transforms/DCPrintDot.cpp:186]:
```suggestion
static DotNode
```

**Review by Dinistro — APPROVED** [review]
Thanks for addressing all the comment. 
LGTM % final nits.

**Comment by luisacicolini** [comment]
Thanks again for your reviews! Everything should be fixed and working now: ready to merge from my side (I have no write permission) :) 
CC: @Dinistro @TaoBi22

**Comment by TaoBi22** [comment]
Cool, I can hit merge for you!
